### PR TITLE
Simple app

### PR DESCRIPTION
**Pull request by macfrei**
I dug up the `Scorekeeper` app Jerry and I used a couple of courses back. I have rewritten it a bit and would like to use it as a base for the component testing session. I would be happy to get feedback. 🙂

### CODE REVIEW

File: .prettierignore [hunk 1]
```diff
@@ -0,0 +1,4 @@
+node_modules
+package-lock.json
+.*
```
Comment (jamarob) [inline .prettierignore:3]: clever you ;)

File: .prettierrc [hunk 2]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (jamarob) [inline .prettierrc:1]: Maybe add prettier as a dev dependency and add a prettify script, not everybody relies on ide extensions for formatting.

File: src/App.js [hunk 3]
```diff
@@ -1,5 +1,79 @@
...
+export default function App() {
+  const [players, setPlayers] = useState([]);
+  const [nameOfGame, setNameOfGame] = useState('');
+  const [currentPage, setCurrentPage] = useState('play');
+  const [history, setHistory] = useState([]);
+
+  return (
+    <AppLayout>
+      <h1>Scorekeeper</h1>
+      {currentPage === 'play' && <GameForm onCreateGame={createGame} />}
```
Comment (jamarob) [inline src/App.js]: Maybe extract the pages to constants?

File: src/App.js [hunk 4]
```diff
@@ -1,5 +1,79 @@
...
+    setPlayers([]);
+    setNameOfGame('');
+    setCurrentPage('play');
+  }
+
+  function resetScores() {
+    setPlayers(players.map(player => ({ ...player, score: 0 })));
+  }
+
+  function handlePlus(index) {
```
Comment (jamarob) [inline src/App.js]: Maybe `increasePlayerScore` and `decreasePlayerScore`?

File: src/App.js [hunk 5]
```diff
@@ -1,5 +1,79 @@
...
+    const currentPlayer = players[index];
+    setPlayers([
+      ...players.slice(0, index),
+      { ...currentPlayer, score: currentPlayer.score + 1 },
+      ...players.slice(index + 1),
+    ]);
```
Comment (jamarob) [inline src/App.js:63]:
```suggestion
    setPlayers(players.map((player,playerIndex) => index === playerIndex ? { ...player, score: player.score + 1 } : player));
```
Imho this is much nicer than the spreading pattern

File: src/components/Button.js [hunk 6]
```diff
@@ -0,0 +1,8 @@
+import styled from 'styled-components';
```
Comment (jamarob) [inline src/components/Button.js:1]:
```suggestion
import styled from 'styled-components/macro';
```

File: src/components/GameForm.js [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+const initialFormData = {
+  nameOfGame: '',
+  playerNames: '',
+};
+
+export default function GameForm({ onCreateGame }) {
+  const [formData, setFormData] = useState(initialFormData);
+
+  return (
+    <FormGrid
```
Comment (jamarob) [inline src/components/GameForm.js]: Maybe just `Form`, I don't care about the layout when I want to understand how it works?

File: src/components/GameForm.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+      aria-labelledby="formHeader"
+      onSubmit={handleSubmit}
+      autoComplete="off"
+    >
+      <h2 id="formHeader">Create a new game</h2>
+      <Input
+        name="nameOfGame"
+        labelText="Name of game"
+        placeholder="e.g. Carcassonne"
```
Comment (jamarob) [inline src/components/GameForm.js]:
```suggestion
        placeholder="e.g. Dodelido"
```
😉

File: src/components/HistoryEntry.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
+import styled from 'styled-components';
+
+export default function HistoryEntry({ nameOfGame, players }) {
+  return (
+    <Wrapper>
+      {nameOfGame}
```
Comment (jamarob) [inline src/components/HistoryEntry.js]: Didn't check the project out to see how it actually looks, but maybe add some style for emphasis of the game name?

File: src/components/HistoryEntry.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+      {players.map((player, index) => (
+        <Player key={index}>
```
Comment (jamarob) [inline src/components/HistoryEntry.js]: Maybe make sure player names are unique or create a unique id for each player on add?

File: src/components/Input.js [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+  name,
+  value,
+  onChange,
+  required,
+}) {
+  return (
+    <>
+      <label htmlFor={name}>{labelText}</label>
+      <InputStyled
+        id={name}
```
Comment (jamarob) [inline src/components/Input.js:15]: I have thought about that in the past a lot, creating a local uuid and assigning it to htmlFor and id will prevent someone from creating 2 inputs with the same id 🤔

**Review by jamarob — APPROVED** [review]
Splendid 🚀 just left some nit picky comments

File: .prettierignore (hunk 1 above)
Comment (macfrei) [reply .prettierignore:3]: Not me, Jerry xD

File: .prettierrc (hunk 2 above)
Comment (macfrei) [reply .prettierrc:1]: True! I will consider it 😎

File: src/App.js (hunk 4 above)
Comment (macfrei) [reply src/App.js]: Yes, good point, thank you!

File: src/App.js (hunk 5 above)
Comment (macfrei) [reply src/App.js:63]: Yes and no, it's a bit shorter but if not used to it is also a bit hard to read, I think. Not sure though what the course will be used to so I would check with Jerry again which syntax should be used.

File: src/components/Button.js (hunk 6 above)
Comment (macfrei) [reply src/components/Button.js:1]: Unfortunately there is a bug with the latest `styled-components` version and `macro` does nothing 😿

File: src/components/GameForm.js (hunk 8 above)
Comment (macfrei) [reply src/components/GameForm.js]: Hahahaha yes! I will change it :)

File: src/components/GameForm.js (hunk 7 above)
Comment (macfrei) [reply src/components/GameForm.js]: Yes, I agree.

File: src/components/HistoryEntry.js (hunk 9 above)
Comment (macfrei) [reply src/components/HistoryEntry.js]: Yep, would be better. Right now it looks a bit `urgh` 😄

File: src/components/HistoryEntry.js (hunk 10 above)
Comment (macfrei) [reply src/components/HistoryEntry.js]: Yes, will do :)

File: src/components/Input.js (hunk 11 above)
Comment (macfrei) [reply src/components/Input.js:15]: Hm maybe, but also, mostly the names are different, no?

File: src/App.js (hunk 3 above)
Comment (macfrei) [reply src/App.js]: How do you mean that exactly?

Comment (jamarob) [reply src/App.js]: `const PLAY="play"` outside of the component, or maybe even export it from another file as it is used in other places too

File: src/App.js (hunk 5 above)
Comment (jamarob) [reply src/App.js:63]: sure! just my preference

File: src/App.js (hunk 3 above)
Comment (macfrei) [reply src/App.js]: Hmm I will think about it 💭 :)

File: src/components/Button.js (hunk 6 above)
Comment (jamarob) [reply src/components/Button.js:1]: 🙀

File: src/components/Input.js (hunk 11 above)
Comment (jamarob) [reply src/components/Input.js:15]: yepp, but names need not be unique for each page. Not sure if my idea is any good, but a good place to start a conversation about it 😅

File: src/App.js (hunk 5 above)
Comment (macfrei) [reply src/App.js:63]: I know :) and it's great to get different preferences and views on topics! :)

File: src/App.js (hunk 3 above)
Comment (badabam) [reply src/App.js]: Good idea, Jan. That way we can prevent errors from typos.

File: src/App.js [hunk 12]
```diff
@@ -1,5 +1,79 @@
...
+      {currentPage === 'play' && <GameForm onCreateGame={createGame} />}
+
+      {currentPage === 'game' && (
+        <GamePage
+          nameOfGame={nameOfGame}
+          players={players}
+          onResetScores={resetScores}
+          onEndGame={endGame}
+          onDecreasePlayerScore={decreasePlayerScore}
+          onIncreasePlayerScore={increasePlayerScore}
```
Comment (badabam) [inline src/App.js:28]: Good naming! Nice!

File: src/App.js (hunk 3 above)
Comment (badabam) [reply src/App.js]: Maybe even create an object
```
const PAGES = {
  PLAY: 'play',
  HISTORY: 'history'
}
```

and then later
```
{ currentPage === PAGES.PLAY && ...
```

File: src/App.js (hunk 5 above)
Comment (badabam) [reply src/App.js:63]: I like both versions. In the new official docs for React, they show `Immer` to keep the updates simple: https://beta.reactjs.org/learn/updating-arrays-in-state#write-concise-update-logic-with-immer. 

That's why we also show it now. With `use-immer`, it is super simple to update such complex states.

File: src/components/GameForm.js [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+  nameOfGame: '',
+  playerNames: '',
+};
+
+export default function GameForm({ onCreateGame }) {
+  const [formData, setFormData] = useState(initialFormData);
+
+  return (
+    <Form
+      aria-labelledby="formHeader"
```
Comment (badabam) [inline src/components/GameForm.js:16]: 👍

File: src/components/Player.js [hunk 14]
```diff
@@ -0,0 +1,41 @@
+import styled from 'styled-components';
+import Button from './Button';
+
+const MAX_HUE_VALUE = 120; // 120 is green
+const MIN_HUE_VALUE = 0; // 0 is red
```
Comment (badabam) [inline src/components/Player.js:5]: 👍

**Review by badabam — APPROVED** [review]
Very nice app!
I added some comments to consider, but that's minor things.

File: src/App.js (hunk 12 above)
Comment (macfrei) [reply src/App.js:28]: All the credit for that goes to @jamarob 🙂

File: src/App.js (hunk 5 above)
Comment (macfrei) [reply src/App.js:63]: I think I would like to leave it as it is - for now. 😬 But thank you for your input - both of you!

**Pull request merged**